### PR TITLE
feat: tipset prechecks

### LINKED ISSUE #1914: Investigate potential vulnerability in ChainMuxer (database flood)
**Issue summary**

As per @aakoshh thread (thank you!) in #fil-forest-dev Slack, this should be investigated further as part of hardening Forest and we should assess the severity and potential solutions.

> [@adlrocha](https://filecoinproject.slack.com/team/U01AML47QJH)
> another thing that appears like vulnerability to me is that once the node enters into Follow mode in the ChainMuxer it will listen to pubsub messages about blocks, and when it gets them it first downloads all the messages via Bitswap, which as a side effect are saved to the database.While this happens, the node doesn't process any other P2P message.Then the full block is validated without its history, the stateless checks I mentioned. And finally if the tipset is not as heavy as our head, it's dropped. Only then do we resume with the next message.It seems like an adversary could send us blocks that cause huge downloads that never get deleted from the IPLD store, of blocks that aren't even connected to the chain. I don't see any penalty, since it could just be considered an old block.(I was just pondering what is safe to be gossipped at all)

**Other information and links**

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:

- adds prechecks (similar to what Lotus does, but more lightweight) before grabbing entire tipsets and putting them into DB. This doesn't 100% mitigate the issue but makes it more difficult to achieve. In our case, it should be _good enough_.

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/1914

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation. All new code adheres to the team's [documentation standards](https://github.com/ChainSafe/forest/wiki/Documentation-practices),
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

### Outside contributions

- [ ] I have read and agree to the [CONTRIBUTING][2] document.
- [ ] I have read and agree to the [AI Policy][3] document. I understand that failure to comply with the guidelines will lead to rejection of the pull request.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md
[2]: https://github.com/ChainSafe/forest/blob/main/CONTRIBUTING.md
[3]: https://github.com/ChainSafe/forest/blob/main/AI_POLICY.md

## Summary by CodeRabbit

* **New Features**
  * Pre-fetch validation for gossip blocks to reject duplicates, bad blocks, out-of-range epochs, and other invalid blocks before fetching.
  * New labeled metric to track rejection reasons for gossip blocks.
  * Reduced redundant processing of duplicate/seen blocks to improve network and persistence efficiency and logging.
* **Tests**
  * Added unit tests covering validation paths and edge cases.

### CODE REVIEW

File: src/chain_sync/validation.rs [hunk 1]
```diff
@@ -145,19 +165,196 @@ impl TipsetValidator<'_> {
...
+
+    fn check_duplicate(
+        cid: Cid,
+        seen_block_cache: &SeenBlockCache,
+    ) -> Result<(), GossipBlockRejectReason> {
+        if seen_block_cache.test_and_insert(&cid) {
+            return Err(GossipBlockRejectReason::DuplicateBlock(cid));
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (LesnyRumcajs) [reply src/chain_sync/validation.rs:257]: Good bot

File: src/chain_sync/bad_block_cache.rs [hunk 2]
```diff
@@ -40,3 +44,30 @@ impl BadBlockCache {
         self.cache.peek_cloned(&(*c).into())
     }
 }
+
+/// Thread-safe LRU cache for tracking recently seen gossip block CIDs.
+/// Used to de-duplicate gossip blocks before expensive message fetching.
+#[derive(Debug)]
```
Comment (hanabi1224) [inline src/chain_sync/bad_block_cache.rs]:
```suggestion
#[derive(Debug, Clone)]
```

File: src/chain_sync/chain_follower.rs [hunk 3]
```diff
@@ -146,6 +147,7 @@ pub async fn chain_follower<DB: Blockstore + Sync + Send + 'static>(
         stateless_mode,
     )));
     let tasks: Arc<Mutex<HashSet<SyncTask>>> = Arc::new(Mutex::new(HashSet::default()));
+    let seen_block_cache = Arc::new(SeenBlockCache::default());
```
Comment (hanabi1224) [inline src/chain_sync/chain_follower.rs]: It's ok to clone `SeenBlockCache` without wrapping it with `Arc`

**Review by hanabi1224 — DISMISSED** [review]

File: src/chain_sync/chain_follower.rs (hunk 3 above)
Comment (LesnyRumcajs) [reply src/chain_sync/chain_follower.rs]: good point

**Review by hanabi1224 — APPROVED** [review]
